### ISSUE TITLE
Error when autoderef of smart pointer occurs in `const fn` is unclear

### ISSUE THREAD

**Issue by SNCPlay42** [rust-lang/rust#91496]
Given the following code: 

```rust
use std::ops::Deref;

struct Foo;

impl Foo {
    const fn foo(&self) {}
}

struct SmartPointer(Foo);

impl Deref for SmartPointer {
    type Target = Foo;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

pub const fn bar() {
    Foo.foo(); // fine
    SmartPointer(Foo).foo(); // fails since we can't constly deref
}
```

The current output is:

```
error[E0015]: calls in constant functions are limited to constant functions, tuple structs and tuple variants
  --> src/lib.rs:21:5
   |
21 |     SmartPointer(Foo).foo(); // fails since we can't constly deref
   |     ^^^^^^^^^^^^^^^^^^^^^^^
```

It's unclear from the diagnostic that the problem is not the call of `foo` itself (which is a `const fn`) but that `SmartPointer` would have to be dereferenced.

Ideally the output should look like:

```
error[E0015]: `SmartPointer` cannot be dereferenced in constant functions
  --> src/lib.rs:21:5
   |
21 |     SmartPointer(Foo).foo(); // fails since we can't constly deref
   |     ^^^^^^^^^^^^^^^^^^^^^^^ dereference occurs due to method call
   |
   = note: operators in constant functions are limited to intrinsic operations
```

Perhaps more generally, the compiler should avoid referring to non-intrinsic operators as "calls" in const errors, or explain that it means calling the operator method. This is *usually* less of a problem because an operator appears in the span.

@rustbot label A-const-fn

**Comment by djc** [comment]
I think this is the same problem:

```rust
match &self.0 {
    Cow::Borrowed(s) => s,
    Cow::Owned(s) => &*s,
}
```

Fails with 

```rust
error[E0015]: calls in constant functions are limited to constant functions, tuple structs and tuple variants
  --> opentelemetry-api/src/common.rs:61:9
   |
61 | /         match &self.0 {
62 | |             Cow::Borrowed(s) => s,
63 | |             Cow::Owned(s) => &*s,
64 | |         }
   | |_________^
```

Despite there clearly being no calls in this function.

**Comment by SNCPlay42** [comment]
Output on current nightly, after #90532:

```
error[E0015]: cannot perform deref coercion on `SmartPointer` in constant functions
  --> src/lib.rs:21:5
   |
21 |     SmartPointer(Foo).foo(); // fails since we can't constly deref
   |     ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: attempting to deref into `Foo`
note: deref defined here
  --> src/lib.rs:12:5
   |
12 |     type Target = Foo;
   |     ^^^^^^^^^^^^^^^^^^
note: impl defined here, but it is not `const`
  --> src/lib.rs:11:1
   |
11 | impl Deref for SmartPointer {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

```

Output from https://github.com/rust-lang/rust/issues/91496#issuecomment-1048556896, (as reproduced like [this](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=97db383dea52309ed26c9c452450b8a5)) has improved as well, although the span is still a bit confusing:

```
error[E0015]: cannot perform deref coercion on `String` in constant functions
    --> src/lib.rs:7:9
     |
7    | /         match &self.0 {
8    | |             Cow::Borrowed(s) => s,
9    | |             Cow::Owned(s) => &*s,
10   | |         }
     | |_________^
     |
     = note: attempting to deref into `str`
note: impl defined here, but it is not `const`
     = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants
```

At any rate, I'm closing this as fixed/in favor of the broader #92380.
